tmt/template.py

```diff
@@ -0,0 +1,109 @@
+TEMPLATE_KEYS = ['story', 'plan', 'test', 'test_metadata']
+
+TemplatesType = dict[str, dict[str, Path]]
+
+
+def _template_folder_exists(path: Path) -> bool:
+    """ Check if given path exists and is a directory """
+    return path.exists() and path.is_dir()
+
+
+def _combine(default: TemplatesType, custom: TemplatesType) -> TemplatesType:
+    """
+    Combines default templates and custom templates.
+    Custom templates have priority and potentially override default templates.
+    """
+    result: TemplatesType = {}
+    for key in default:
+        result[key] = {**default[key], **custom.get(key, {})}
+    for key in custom:
+        if key not in result:
+            result[key] = custom[key]
+    return result
+
+
+def _get_template_file_paths(path: Path) -> dict[str, Path]:
+    """ Return a dictionary of template file paths """
+    return {file.name: file for file in path.iterdir() if file.is_file()}
+
+
+def _get_templates(root_dir: Path) -> TemplatesType:
+    """ Find all templates in given root directory """
+    templates: TemplatesType = {}
+    for key in TEMPLATE_KEYS:
+        templates_dir = root_dir / key
+        if _template_folder_exists(templates_dir):
+            template_files = _get_template_file_paths(templates_dir)
+            if template_files:
+                templates[key] = template_files
+    return templates
+
+
+class Template:
+    def __init__(self, custom_template_path: Optional[Path] = None):
+        self.custom_template_path = custom_template_path or DEFAULT_CUSTOM_TEMPLATES_PATH
+        self._init_custom_templates_folder()
+        self._environment = tmt.utils.default_template_environment()
+
+    @cached_property
+    def templates(self) -> TemplatesType:
+        """ Return all available templates (default and optional) """
+        return _combine(self.default_templates, self.custom_templates)
+
+    @cached_property
+    def default_templates(self) -> TemplatesType:
+        templates_dir = tmt.utils.resource_files('templates/')
+        templates = _get_templates(templates_dir)
+        if not templates:
+            raise tmt.utils.GeneralError(f"Could not find default templates in '{templates_dir}'.")
+        return templates
+
+    @cached_property
+    def custom_templates(self) -> TemplatesType:
+        return _get_templates(self.custom_template_path)
+
+    def render_default_plan(self) -> str:
+        """ Return default plan template """
+        try:
+            path = self.default_templates['default']['plan']
+        except KeyError:
+            raise tmt.utils.GeneralError("Default plan template not found.")
+
+        return self.render_file(path, plan_name=DEFAULT_PLAN_NAME)
+
+    def render_from_url(self, url: str, **variables: Any) -> str:
+        """ Render template from given url """
+        template = tmt.utils.get_file_content_from_url(url)
+        return self.render(template, **variables)
+
+    def render(self, template_str: str, **variables: Any) -> str:
```

Comment:
This one looks like not being used by external code, would it make sense to mark as "private", `_render()`? It seems `render_from_url` and `render_file` only are called from outside.

Reply:
Marked as private in 78d7217caecb8c3bd05d92445264de9de4bfaf02.